File: scripts/VerifyDependencyBoundaries.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.stream.Stream;
import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilderFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;

/** Verifies Maven dependency ownership against JFoundry module boundaries. */
public final class VerifyDependencyBoundaries {

    private VerifyDependencyBoundaries() {
    }

    public static void main(String[] args) throws Exception {
        Path root = args.length == 0 ? Path.of(".") : Path.of(args[0]);
        List<String> violations = new ArrayList<>();
        try (Stream<Path> paths = Files.walk(root)) {
            paths.filter(path -> path.getFileName().toString().equals("pom.xml"))
                    .filter(path -> !containsExcludedDirectory(root.relativize(path)))
                    .forEach(path -> inspect(root, path, violations));
        }

        if (!violations.isEmpty()) {
            violations.forEach(System.err::println);
            System.err.println("Dependency boundary check failed with " + violations.size() + " violation(s).");
            System.exit(1);
        }
    }

    private static void inspect(Path root, Path pom, List<String> violations) {
        try {
            Document document = parse(pom);
            ModuleKind module = classify(root.relativize(pom.getParent()));
            for (Element dependencies : elements(document.getDocumentElement(), "dependencies")) {
                Element parent = parentElement(dependencies);
                if (parent != null && "plugin".equals(localName(parent))) {
                    continue;
                }
                String context = hasAncestor(dependencies, "dependencyManagement")
                        ? "dependencyManagement"
                        : "dependencies";
                for (Element dependency : childElements(dependencies, "dependency")) {
                    String groupId = text(dependency, "groupId");
                    String artifactId = text(dependency, "artifactId");
                    if (groupId.isBlank() || artifactId.isBlank()) {
                        continue;
                    }
                    String scope = text(dependency, "scope");
                    check(module, groupId, artifactId, context + (scope.isBlank() ? "" : ":" + scope),
                            root.relativize(pom), violations);
                }
            }
        } catch (Exception exception) {
            violations.add(root.relativize(pom) + " [parse-error] " + exception.getMessage());
        }
    }

    private static void check(ModuleKind module, String groupId, String artifactId, String context, Path pom,
            List<String> violations) {
        String coordinate = (groupId + ":" + artifactId).toLowerCase(Locale.ROOT);
        boolean spring = isSpring(coordinate);
        boolean quarkus = isQuarkus(coordinate);
        boolean helidon = isHelidon(coordinate);

        if (module == ModuleKind.CORE && (spring || quarkus || helidon) && !isAllowedCoreApi(groupId, artifactId)) {
            violations.add(format(pom, coordinate, context, "core-runtime-dependency"));
        }
        if (module == ModuleKind.FOUNDATION_BOM && (spring || quarkus || helidon || isRuntimeMarker(coordinate))) {
            violations.add(format(pom, coordinate, context, "foundation-runtime-coordinate"));
        }
        if (module == ModuleKind.JAKARTA && (spring || quarkus || helidon)) {
            violations.add(format(pom, coordinate, context, "jakarta-cross-runtime-dependency"));
        }
        if ((module == ModuleKind.SPRING || module == ModuleKind.SPRING_BOM) && (quarkus || helidon)) {
            violations.add(format(pom, coordinate, context, "spring-cross-runtime-dependency"));
        }
        if ((module == ModuleKind.QUARKUS || module == ModuleKind.QUARKUS_BOM) && (spring || helidon)) {
            violations.add(format(pom, coordinate, context, "quarkus-cross-runtime-dependency"));
        }
        if ((module == ModuleKind.HELIDON || module == ModuleKind.HELIDON_BOM) && (spring || quarkus)) {
            violations.add(format(pom, coordinate, context, "helidon-cross-runtime-dependency"));
        }
        if (isRuntimeBom(module) && isDisallowedRuntimeBomImport(coordinate)) {
            violations.add(format(pom, coordinate, context, "runtime-bom-import"));
        }
    }

    private static String format(Path pom, String coordinate, String context, String rule) {
        return pom + " [" + rule + "] " + coordinate + " (" + context + ")";
    }

    private static boolean isAllowedCoreApi(String groupId, String artifactId) {
        return groupId.equals("jakarta.persistence") && artifactId.equals("jakarta.persistence-api");
    }

    private static boolean isSpring(String coordinate) {
        return coordinate.startsWith("org.springframework:")
                || coordinate.startsWith("org.springframework.")
                || coordinate.contains(":spring-")
                || coordinate.contains("-spring-")
                || coordinate.contains(":jmolecules-spring")
                || coordinate.contains(":jfoundry-" ) && coordinate.contains("-spring");
    }

    private static boolean isQuarkus(String coordinate) {
        return coordinate.startsWith("io.quarkus:")
                || coordinate.startsWith("io.quarkus.")
                || coordinate.contains(":jfoundry-" ) && coordinate.contains("-quarkus");
    }

    private static boolean isHelidon(String coordinate) {
        return coordinate.startsWith("io.helidon:")
                || coordinate.startsWith("io.helidon.")
                || coordinate.contains(":jfoundry-" ) && coordinate.contains("-helidon");
    }

    private static boolean isRuntimeMarker(String coordinate) {
        return coordinate.contains("-deployment") || coordinate.contains("-starter");
    }

    private static boolean isRuntimeBom(ModuleKind module) {
        return module == ModuleKind.SPRING_BOM
                || module == ModuleKind.QUARKUS_BOM
                || module == ModuleKind.HELIDON_BOM;
    }

    private static boolean isDisallowedRuntimeBomImport(String coordinate) {
        return coordinate.equals("io.github.xfoundries:jfoundry-dependencies")
                || coordinate.equals("io.github.xfoundries:jfoundry-foundation-dependencies")
                || coordinate.equals("io.github.xfoundries:jfoundry-spring-boot-dependencies")
                || coordinate.equals("io.github.xfoundries:jfoundry-spring-cloud-dependencies")
                || coordinate.equals("io.github.xfoundries:jfoundry-quarkus-dependencies")
                || coordinate.equals("io.github.xfoundries:jfoundry-helidon-dependencies");
    }

    private static ModuleKind classify(Path relativePath) {
        String path = relativePath.toString().replace('\\', '/');
        if (path.startsWith("jfoundry-core/")) {
            return ModuleKind.CORE;
        }
        if (path.startsWith("jfoundry-boms/jfoundry-spring-") && path.endsWith("-dependencies")) {
            return ModuleKind.SPRING_BOM;
        }
        if (path.startsWith("jfoundry-boms/jfoundry-quarkus-dependencies")) {
            return ModuleKind.QUARKUS_BOM;
        }
        if (path.startsWith("jfoundry-boms/jfoundry-helidon-dependencies")) {
            return ModuleKind.HELIDON_BOM;
        }
        if (path.startsWith("jfoundry-runtime/jfoundry-spring/")) {
            return ModuleKind.SPRING;
        }
        if (path.startsWith("jfoundry-runtime/jfoundry-jakarta/")) {
            return ModuleKind.JAKARTA;
        }
        if (path.startsWith("jfoundry-runtime/jfoundry-quarkus/")) {
            return ModuleKind.QUARKUS;
        }
        if (path.startsWith("jfoundry-runtime/jfoundry-helidon/")) {
            return ModuleKind.HELIDON;
        }
        if (path.startsWith("jfoundry-boms/jfoundry-foundation-dependencies")) {
            return ModuleKind.FOUNDATION_BOM;
        }
        return ModuleKind.OTHER;
    }

    private static boolean containsExcludedDirectory(Path relativePath) {
        for (Path part : relativePath) {
            if (part.toString().equals("target") || part.toString().equals(".git") || part.toString().equals(".worktrees")) {
                return true;
            }
        }
        return false;
    }

    private static Document parse(Path pom) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        factory.setNamespaceAware(true);
        factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
        return factory.newDocumentBuilder().parse(pom.toFile());
    }

    private static List<Element> elements(Element root, String name) {
        List<Element> result = new ArrayList<>();
        collect(root, name, result);
        return result;
    }

    private static void collect(Element element, String name, List<Element> result) {
        if (name.equals(localName(element))) {
            result.add(element);
        }
        for (Node node = element.getFirstChild(); node != null; node = node.getNextSibling()) {
            if (node instanceof Element child) {
                collect(child, name, result);
            }
        }
    }

    private static List<Element> childElements(Element parent, String name) {
        List<Element> result = new ArrayList<>();
        for (Node node = parent.getFirstChild(); node != null; node = node.getNextSibling()) {
            if (node instanceof Element child && name.equals(localName(child))) {
                result.add(child);
            }
        }
        return result;
    }

    private static Element parentElement(Element element) {
        return element.getParentNode() instanceof Element parent ? parent : null;
    }

    private static boolean hasAncestor(Element element, String name) {
        Node node = element.getParentNode();
        while (node instanceof Element parent) {
            if (name.equals(localName(parent))) {
                return true;
            }
            node = parent.getParentNode();
        }
        return false;
    }

    private static String localName(Element element) {
        return element.getLocalName() == null ? element.getNodeName() : element.getLocalName();
    }

    private static String text(Element parent, String name) {
        for (Element child : childElements(parent, name)) {
            return child.getTextContent().trim();
        }
        return "";
    }

    private enum ModuleKind {
        CORE, FOUNDATION_BOM, JAKARTA, SPRING, QUARKUS, HELIDON,
        SPRING_BOM, QUARKUS_BOM, HELIDON_BOM, OTHER
    }
}

File: scripts/VerifyDependencyBoundaries.rb
require "find"
require "pathname"
require "rexml/document"
require "yaml"

ROOT = Pathname(ARGV.fetch(0, ".")).expand_path
EXCLUDED_DIRECTORIES = %w[.git .worktrees target].freeze

def classify(relative_path)
  path = relative_path.to_s.tr("\\", "/")
  return :core if path.start_with?("jfoundry-core/")
  return :spring_bom if path.start_with?("jfoundry-boms/jfoundry-spring-") && path.end_with?("-dependencies")
  return :quarkus_bom if path.start_with?("jfoundry-boms/jfoundry-quarkus-dependencies")
  return :helidon_bom if path.start_with?("jfoundry-boms/jfoundry-helidon-dependencies")
  return :spring if path.start_with?("jfoundry-runtime/jfoundry-spring/")
  return :jakarta if path.start_with?("jfoundry-runtime/jfoundry-jakarta/")
  return :quarkus if path.start_with?("jfoundry-runtime/jfoundry-quarkus/")
  return :helidon if path.start_with?("jfoundry-runtime/jfoundry-helidon/")
  return :foundation_bom if path.start_with?("jfoundry-boms/jfoundry-foundation-dependencies")

  :other
end

def spring?(coordinate)
  coordinate.start_with?("org.springframework:", "org.springframework.") ||
    coordinate.include?(":spring-") || coordinate.include?("-spring-") ||
    coordinate.include?(":jmolecules-spring") ||
    (coordinate.include?(":jfoundry-") && coordinate.include?("-spring"))
end

def quarkus?(coordinate)
  coordinate.start_with?("io.quarkus:", "io.quarkus.") ||
    (coordinate.include?(":jfoundry-") && coordinate.include?("-quarkus"))
end

def helidon?(coordinate)
  coordinate.start_with?("io.helidon:", "io.helidon.") ||
    (coordinate.include?(":jfoundry-") && coordinate.include?("-helidon"))
end

def runtime_bom?(module_kind)
  %i[spring_bom quarkus_bom helidon_bom].include?(module_kind)
end

def disallowed_runtime_bom_import?(coordinate)
  %w[
    io.github.xfoundries:jfoundry-dependencies
    io.github.xfoundries:jfoundry-foundation-dependencies
    io.github.xfoundries:jfoundry-spring-boot-dependencies
    io.github.xfoundries:jfoundry-spring-cloud-dependencies
    io.github.xfoundries:jfoundry-quarkus-dependencies
    io.github.xfoundries:jfoundry-helidon-dependencies
  ].include?(coordinate)
end

def check_dependency(module_kind, dependency, context, pom, violations)
  group_id = dependency.fetch("groupId", "").to_s
  artifact_id = dependency.fetch("artifactId", "").to_s
  return if group_id.empty? || artifact_id.empty?

  coordinate = "#{group_id}:#{artifact_id}".downcase
  scope = dependency.fetch("scope", "").to_s
  location = scope.empty? ? context : "#{context}:#{scope}"
  spring = spring?(coordinate)
  quarkus = quarkus?(coordinate)
  helidon = helidon?(coordinate)
  rules = []
  if module_kind == :core && (spring || quarkus || helidon) &&
      coordinate != "jakarta.persistence:jakarta.persistence-api"
    rules << "core-runtime-dependency"
  end
  if module_kind == :foundation_bom &&
      (spring || quarkus || helidon || coordinate.include?("-deployment") || coordinate.include?("-starter"))
    rules << "foundation-runtime-coordinate"
  end
  rules << "jakarta-cross-runtime-dependency" if module_kind == :jakarta && (spring || quarkus || helidon)
  rules << "spring-cross-runtime-dependency" if %i[spring spring_bom].include?(module_kind) && (quarkus || helidon)
  rules << "quarkus-cross-runtime-dependency" if %i[quarkus quarkus_bom].include?(module_kind) && (spring || helidon)
  rules << "helidon-cross-runtime-dependency" if %i[helidon helidon_bom].include?(module_kind) && (spring || quarkus)
  if runtime_bom?(module_kind) && disallowed_runtime_bom_import?(coordinate)
    rules << "runtime-bom-import"
  end
  rules.each { |rule| violations << "#{pom} [#{rule}] #{coordinate} (#{location})" }
end

def yaml_dependency_sets(model)
  sets = []
  dependencies = model["dependencies"]
  sets << [dependencies, "dependencies"] if dependencies.is_a?(Array)
  managed = model.dig("dependencyManagement", "dependencies")
  sets << [managed, "dependencyManagement"] if managed.is_a?(Array)
  Array(model["profiles"]).each do |profile|
    sets.concat(yaml_dependency_sets(profile)) if profile.is_a?(Hash)
  end
  sets
end

def xml_dependency_sets(document)
  sets = []
  REXML::XPath.each(document, "//*[local-name()='dependencies']") do |dependencies|
    next if dependencies.parent&.name == "plugin"

    context = REXML::XPath.first(dependencies, "ancestor::*[local-name()='dependencyManagement']") ?
      "dependencyManagement" : "dependencies"
    values = dependencies.get_elements("*[local-name()='dependency']").map do |dependency|
      %w[groupId artifactId scope].to_h do |name|
        element = dependency.elements["*[local-name()='#{name}']"]
        [name, element&.text.to_s.strip]
      end
    end
    sets << [values, context]
  end
  sets
end

def load_dependency_sets(pom)
  if pom.extname == ".yaml"
    model = YAML.safe_load(pom.read, permitted_classes: [], permitted_symbols: [], aliases: true)
    raise "top-level model must be a mapping" unless model.is_a?(Hash)

    yaml_dependency_sets(model)
  else
    xml_dependency_sets(REXML::Document.new(pom.read))
  end
end

violations = []
Find.find(ROOT.to_s) do |entry|
  path = Pathname(entry)
  if path.directory? && path != ROOT && EXCLUDED_DIRECTORIES.include?(path.basename.to_s)
    Find.prune
  end
  next unless path.file? && %w[pom.yaml pom.xml].include?(path.basename.to_s)

  relative_pom = path.relative_path_from(ROOT)
  module_kind = classify(relative_pom.dirname)
  begin
    load_dependency_sets(path).each do |dependencies, context|
      dependencies.each do |dependency|
        check_dependency(module_kind, dependency, context, relative_pom, violations)
      end
    end
  rescue StandardError => error
    violations << "#{relative_pom} [parse-error] #{error.message}"
  end
end

unless violations.empty?
  violations.each { |violation| warn violation }
  warn "Dependency boundary check failed with #{violations.size} violation(s)."
  exit 1
end
